File: mergesort/mergesort.cpp
#include <iostream>
#include <tuple>
#include <vector>

using namespace std;

tuple<vector<int>, vector<int>> halfs(vector<int> xs) {
    if (xs.empty()) return {{}, {}};
    vector<int> ys, zs;
    int mid = xs.size() / 2;
    for (int i = 0; i < xs.size(); i++) {
        if (i < mid) {
            ys.push_back(xs[i]);
        } else {
            zs.push_back(xs[i]);
        }
    }
    return {ys, zs};
}

vector<int> merge(vector<int> xs, vector<int> ys) {
    vector<int> zs(xs.size() + ys.size());
    int ix = 0;
    int iy = 0;
    while (ix + iy < zs.size()) {
        if (ix == xs.size()) {
            while (iy < ys.size()) {
                zs[ix + iy] = ys[iy];
                iy ++;
            }
        } else if (iy == ys.size()) {
            while (ix < xs.size()) {
                zs[ix + iy] = xs[ix];
                ix ++;
            }
        } else if (xs[ix] < ys[iy]) {
            zs[ix + iy] = xs[ix];
            ix ++;
        } else {
            zs[ix + iy] = ys[iy];
            iy ++;
        }
    }
    return zs;
}

vector<int> mergesort(vector<int> xs) {
    if (xs.size() < 2) return xs;
    vector<int> ys, zs;
    tie(ys, zs) = halfs(xs);
    return merge(mergesort(ys), mergesort(zs));
}

void print(const vector<int> &xs) {
    for (int x : xs) {
        cout << x << " ";
    }
    cout << endl;
}

int main() {
    int n;
    vector<int> xs;

    while (cin >> n) {
        xs.push_back(n);
    }

    xs = mergesort(xs);
    print(xs);

    return 0;
}

File: mergesort/mergesort.hs
-- Author: Fernando de Jesús Soto

mergesort :: Ord a => [a] -> [a]
mergesort [] = []
mergesort [x] = [x]
mergesort xs = merge (mergesort xs')  (mergesort xs'')
    where (xs', xs'') = halfs xs

merge :: Ord a => [a] -> [a] -> [a]
merge [] ys = ys
merge xs [] = xs
merge (x:xs) (y:ys)
    | x < y = x : merge xs (y:ys)
    | otherwise = y : merge (x:xs) ys

halfs :: [a] -> ([a], [a])
halfs [] = ([], [])
halfs xs = (take mid xs, drop mid xs)
    where mid = length xs `div` 2
